### PR TITLE
Milestone 5 - Create Friendship Relationship

### PR DESCRIPTION
**Pull request by guirecordon**
### Create a model with associations and other features for friendships

- created a friendships table
- added the 'belongs_to' relationships to the friendship model
- added the 'had_many' relationships to the user model
- created a friends method in the user model to account for the _inverse_ friendships
- created _pending_friends_, _friend_requests_, and _confirm_friends_ helper methods inside of the user model
- created unit tests for all the models
- created integration tests for the user registration, login and posts.

We ran Rubocop and the Linters and found no problems. We're thus submitting it for review.

Thank you.
Gui/
Jocelyn

### CODE REVIEW

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great start @team :clap: However, some requirements need to be implemented in order to get approval for this milestone. Check :arrow_down: 

#### Behaviour (required) :warning: 
* [ ] User can invite friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/send_invitation.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [ ] User can see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [ ] User has to accept friendship request to start friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [ ] User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [ ] There is a button to invite user to friendship on users list page and on single user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [ ] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
#### Implementation (required) :warning: 
* [ ] Use scope, ActiveRecord method or SQL queries to retrieve record. Avoid usage of enumerable methods.
* [ ] Clicking on the link to visit the profile of a user triggers a memory stack error. Kindly fix it.
#### Views & Helpers (required) :warning: 

* [ ] There should be NO application logic in the views nor helpers (no `Post.new`, `Comment.create`, `current_user.friends.find_by(...)`...). Everything connected with data processing should be encapsulated in models and used in controllers.
* [ ] View-related logic can be encapsulated in View helpers (app/helpers/\*_helper.rb)
* [ ] There shoud be absolutely NO logic in views.

  > Views are only for displaying information that became available to it from its corresponding controller not generating it. **Resources**
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright
### That's it :relieved: When finished, kindly submit a code re-review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Comment by guirecordon** [comment]
Hi, we believe some of the requested changes are outside the scope of this milestone. We are thus requesting for a new review.

Thank you.

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

#### Hello Team!, Great job! 👏 

You've done an awesome job so far 👌. But there are still some changes that are necessary for this project to be approved
.
- [ ] Clicking on all users throws an error, kindly fix this: 
Screen Shot 2020-11-10 at 8 17 35 PM

- [ ] Clicking on my username throws an error as well, kindly fix this: 
Screen Shot 2020-11-10 at 8 22 05 PM

### Kindly Note that it's part of the requirement of this milestone for you to achieve the following

- [ ]  User can invite friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/send_invitation.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
 
- [ ] User can see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

- [ ]  User should be able to accept friendship request to start a friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

 - [ ] User should be able to reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

- [ ]  There should be a button to invite a user to friendship on users list page and on single-user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

 - [ ] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

 - [ ] Add tests! ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/tests.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

Having difficulties? 
Check out this article as a hint: [https://smartfunnycool.com/friendships-in-activerecord/](https://smartfunnycool.com/friendships-in-activerecord/.)

- [ ] There should be NO application logic in the views, so nothing like `if` or `unless`. View-related logic should be encapsulated in View helpers (app/helpers/*_helper.rb). [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright
These articles may help: [Where-Do-I-Put-My-Code](http://codefol.io/posts/Where-Do-I-Put-My-Code/) and [rails-view-helpers-best-practices-guide](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4)

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Comment by Jmagero** [comment]
### Requested changes
- [x] Clicking on all users throws an error
- [x] Clicking on my username throws an error as well, kindly fix this:
- [x] User can invite friends 
- [x] User can see friendships requests
- [x] User should be able to accept friendship request to start a friendship
- [x] User should be able to reject a friendship request- **Requesting guidance** 
- [x]  There should be a button to invite a user to friendship on users list page and on the single-user page
- [x]  Posts timeline includes posts written by logged-in user and their friends
- [x]  Add tests! 
- [x] There should be NO application logic in the views, so nothing like if or unless

File: app/models/friendship.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class Friendship < ApplicationRecord
+  belongs_to :user
+  belongs_to :friend, class_name: 'User'
+
+  # validates_presence_of :user_id
+  # validates_presence_of :friend_id
+  # validates_presence_of :confirmed
```
Comment (ghost) [inline app/models/friendship.rb]: Remove the commented code.

File: spec/models/friendship_spec.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  context 'validations' do
+    it 'creates friendship when user accepts invitation' do
+      f = Friendship.new(user_id: user1.id, friend_id: user2.id, confirmed: true)
+      expect(f).to be_valid
+    end
+
+    # it 'doesn\'t create frienshitp when user does not accept invitation' do
+    #   f = Friendship.new(user_id: user1.id, friend_id: user2.id, confirmed: false)
+    #   expect(f).not_to be_valid
+    # end
```
Comment (ghost) [inline spec/models/friendship_spec.rb]: Remove the commented code.

File: app/models/friendship.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class Friendship < ApplicationRecord
+  belongs_to :user
+  belongs_to :friend, class_name: 'User'
```
Comment (ghost) [inline app/models/friendship.rb:3]: The usage of `class_name` has not been tested in your association test. Hint: Refer to [shoulda matchers ](http://matchers.shoulda.io/docs/v4.0.1/Shoulda/Matchers/ActiveRecord.html#belong_to-instance_method) on this.

File: spec/models/post_spec.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  context 'testing associations Post' do
+    it 'has many likes' do
+      x = Post.reflect_on_association(:likes)
+      expect(x.macro).to eq(:has_many)
+    end
+
+    it 'has many comments' do
+      x = Post.reflect_on_association(:comments)
+      expect(x.macro).to eq(:has_many)
+    end
```
Comment (ghost) [inline spec/models/post_spec.rb:32]: The usage of `dependent` has not been tested in your association test. Hint: Refer to [shoulda matchers ](http://matchers.shoulda.io/docs/v4.0.1/Shoulda/Matchers/ActiveRecord.html#belong_to-instance_method) on this.

File: app/models/user.rb [hunk 5]
```diff
@@ -5,8 +5,40 @@ class User < ApplicationRecord
          :recoverable, :rememberable, :validatable
 
   validates :name, presence: true, length: { maximum: 20 }
+  validates :password, presence: true, length: { minimum: 6 }
+  validates :password_confirmation, presence: true, confirmation: { case_sensitive: true }
 
-  has_many :posts
+  has_many :posts, dependent: :destroy
   has_many :comments, dependent: :destroy
   has_many :likes, dependent: :destroy
```
Comment (ghost) [inline app/models/user.rb:13]: The usage of `dependent` has not been tested in your association test. Hint: Refer to [shoulda matchers ](http://matchers.shoulda.io/docs/v4.0.1/Shoulda/Matchers/ActiveRecord.html#belong_to-instance_method) on this.

File: app/models/user.rb [hunk 6]
```diff
@@ -5,8 +5,40 @@ class User < ApplicationRecord
...
+
+  has_many :friendships
+  has_many :inverse_friendships, class_name: 'Friendship', foreign_key: 'friend_id'
```
Comment (ghost) [inline app/models/user.rb:16]: The usage of `class_name` and `foreign_key` has not been tested in your association test. Hint: Refer to [shoulda matchers ](http://matchers.shoulda.io/docs/v4.0.1/Shoulda/Matchers/ActiveRecord.html#belong_to-instance_method) on this.

File: app/models/user.rb [hunk 7]
```diff
@@ -5,8 +5,40 @@ class User < ApplicationRecord
...
+  def confirm_friend(user)
+    friendship = inverse_friendships.find { |friend| friend.user == user }
+    friendship.confirmed = true
+    friendship.save
+  end
+
+  def friend?(user)
+    friends.include?(user)
+  end
 end
```
Comment (ghost) [inline app/models/user.rb]: Add tests for the above methods as well. If you are not using any of them outside the class, you can declare them private and not have to write tests for them.

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required 🛑

## Hello team 👋,

**Great job implementing the requested changes 👏. However, there are still a few changes needed.**

1). **A user should not be able to see another user's post unless they are confirmed friends. See [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from pathwright.**
2). **After a user sends a friend request, the friend request button should be disabled and converted to a paragraph which informs that the request has been sent. See an example in the screenshot below.**

 before a request is sent  | After a request is sent
 ---------------------------| ------------------------
before_request | after_request

3). **When I press the reject friend request button, I get an error.**
delete_request

4). **Add tests for your helpers, and system tests for the following:**

- when a user likes a post
- when a user comments on a post.
- when a user sends a friend request.
- when a user accepts and rejects a friend request.

5). **There is logic (if statements and each loop) in the following views files:**

```
- layouts/application.html.erb.
- posts/_post.html.erb
- users/index.html.erb
- users/show.html.erb
```
 View-related logic should be encapsulated in View helpers `(app/helpers/*_helper.rb)`. Hint: Check out this [article](https://dev.to/saintaze/rendering-list-items-in-ruby-on-rails-6-48ae) on how to represent list items in `views`.

6). **See more in the comments below.**

## Happy coding 🥂.

**PS: You can reach me on slack (Rose Wanjohi).**

**Comment by Jmagero** [comment]
**Fix Required changes**

- [X] User should not be able to see another user's post unless they are confirmed, friends. 
- [X] After a user sends a friend request, the friend request button should be disabled and converted to a paragraph that informs that the request has been sent. 
- [x] When I press the reject friend request button, I get an error.
- [x] Add tests for your helpers, and system tests.
- [x] Remove logic in the view files

File: app/views/layouts/application.html.erb [hunk 8]
```diff
@@ -1,14 +1,22 @@
...
     <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
   </head>
   <body>
     <nav class="nav">
+    <% if logged_in? %>	      
+      <div class="nav-profile"> 	      
+        <%= link_to (gravatar_for current_user), user_path(current_user) %>	
+        <%= link_to current_user.name, user_path(current_user), id: 'logo'%>	
+      </div>	
+    <% end %>
```
Comment (acushlakoncept) [inline app/views/layouts/application.html.erb]: - [ ] There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  

**The Resources below may come in handy**
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

File: app/helpers/application_helper.rb [hunk 9]
```diff
@@ -15,4 +19,19 @@ def like_or_dislike_btn(post)
...
+  #   end
+  # end
+
+  # def is_current_user
+  #   if current_user?
+  #     link_to 'Sign out', destroy_user_session_path, method: :delete
+  #   else
+  #     link_to 'Sign in', user_session_path
+  #   end
+  # end
```
Comment (acushlakoncept) [inline app/helpers/application_helper.rb]: It’s better to delete commented piece of codes. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Excellent work! the project works perfectly well and great job on your specs
Let us quickly add some finishing touches and request another review.

- [ ] With this level of progress, an update on the readme file would be perfectly in order.

I have made a few comments on our codebase, kindly take a look and fix the requested changes

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Comment by Jmagero** [comment]
**Requested Changes**

- [x] Update READMe
- [X] Removed logic from the view file

**Review by arslanbisharat — APPROVED** [review]
# Approved :smile: 

Hello Developer :wave:

I just reviewed your project, and I am thrilled to inform you that your project is approved.
Congratulations on achieving this milestone in your career. :clap:

- [ ] You have implemented all the changes requested by previous TSEs. Good Job!
- [ ] You have a nice PR and Commit history including Readme. 
- [ ] ALL linters passing. Nice work!
_I am looking forward to seeing your next project._

Stay safe and give me a smile 😇

### Happy Coding
------------------------------------------------------------------------------------------------
**Regards,
Arslan Bisharat
slack: arslanbisharat**

**Pull request merged**